create named sitemaps

ElectricMaxxx opened the pull request:
fix #221, fix #212.

Introduce named sitemaps with its own defintions of default_change_frequency (later extracted by an guesser - just as fallback) and the templates.

Following tasks will be:
* [x] named sitemaps with configuration and schema
* [x] create a voters wich reduce the list of documents, based on the result of the provider
* [x] let guesser do the work of creating url information baed on the given documents (after vote)
* [x] each of them will be taged services with an attribute `sitemap` to define different for each sitemap
* [x] fix travis failing test
* [x] autocreate defautl sitemape configuration

Doc issue: https://github.com/symfony-cmf/symfony-cmf-docs/issues/681

ElectricMaxxx commented on Tests/Resources/Fixtures/config/config2.xml:
```diff
@@ -11,5 +11,12 @@
+    <sitemap enabled="true">
+        <configuration default-chan-frequency="never" enabled="true" name="default">
+            <templates>
+                <template format="xml" template="test.xml" />
```
Not sure if the nested way is the right one. It was the easiest way of defining a schema and it is the way i expect the values in the `CmfSeoExtension`

ElectricMaxxx commented:
Or should i do all tasks in this PR? then i would change the list above into checkboxes.

dbu commented on DependencyInjection/Configuration.php:
```diff
@@ -89,7 +89,28 @@ public function getConfigTreeBuilder()
+
+                    ->end()
+                ->end()
+                ->arrayNode('sitemap')
```
i think you duplicated the sitemap arrayNode here

dbu commented on Controller/SitemapController.php:
```diff
@@ -78,8 +86,8 @@ public function indexAction($_format)
         }
 
         $urls = $this->urlProvider->getUrlInformation();
```
i would put the urlProvider into the configuration as well. then i can have one provider for each sitemap.

dbu commented on DependencyInjection/CmfSeoExtension.php:
```diff
@@ -233,9 +233,18 @@ private function loadSitemapHandling($config, XmlFileLoader $loader, ContainerBu
+        }
+
+        // set a base default change frequency by now todo change that when votes respect sitemap by name
+        if (isset($config['configurations']['default']['default_change_frequency'])) {
```
should we autocreate the default sitemap if its not explicitly defined? 

and how do we handle default templates? we really need #217 now but i would say: if no templates are defined, we add the default ones for html and xml, otherwise we don't do anything extra and assume the user did what he wanted.

ElectricMaxxx replied:
this is a task i didn't get yesterday, found no real example tried to look into doctrine-bundle.

dbu replied:
i think the best approach is to use the Configuration beforeValidation callbacks for this, not sure if it would even be possible in pure declarative code. i can have a go at that tomorrow if you want

dbu commented:
i would prefer to do this in several PR so that we merge as soon as one part is good.

ElectricMaxxx commented:
You have had your chance, its to Late now. Got almost everything except the voter working in an half an Hour. :-)

dbu commented:
do you ever sleep? great job!

dbu commented on Tests/Resources/app/config/cmf_seo.yml:
```diff
@@ -14,7 +14,13 @@ cmf_seo:
+                default_change_frequency: never
+                templates:
+                    xml: 'CmfSeoBundle:Sitemap:index.xml.twig'
+                    html: 'CmfSeoBundle:Sitemap:index.html.twig'
```
imho we should not need this explicitly when its the default values. we should just say sitemap: true and the default sitemap with these defaults is automatically defined. the Configuration could auto-add things. 

i can do a PR tomorrow to do this if you want.

ElectricMaxxx replied:
:+1:

ElectricMaxxx commented:
So ... This will fail for sure, as i still miss the voters. I need to do some refactoring in code duplications and re-write the tests for be base provider and guesser. But it should show the way it can work.

ElectricMaxxx commented on Controller/SitemapController.php:
```diff
@@ -77,12 +94,19 @@ public function indexAction($_format)
+        $urlInformations = array();
+        $documents = $this->routesForSitemapProvider->getDocumentsForSitemap($sitemap);
+        foreach ($documents as $document) {
+            $urlInformation = new UrlInformation();
```
I would implement the blocker of the voter right here.

ElectricMaxxx commented:
@dbu i hope you understand the mechanism i tried to implement. Both, the guesser and the provider, are tagged services with an attribute for the priority and the sitemap name. Priority defaults to 0, the name to `default` at all. The providers are responsible to provide a list of documents/objects for the sitemap. The guesser will get each of them and change a given url information object. As botth, guesser and providers a priotisised you can chain them together in that way you want and can be bound to a specific sitemap. Thats the reason for both chains which implementing the same interface as the guesser or providers.

dbu commented:
sorry, i won't get much done today. but i looked over this and want to propose we do things a bit differently:

* the controller now became huge and is bound to the documentprovider - guesser paradigm
* i suggest that we keep the UrlInformationProvider as it was, and have an implementation with the code you are now adding to the controller method.
* i would call the thing you call document provider a "document loader".

ElectricMaxxx commented:
OK, pushing it to an extra service sound sane, so :+1:

Just the name for the loader should be something else. Cause what about 
simple objects somebody else is getting from the ORM?

So what about an ContentLoader?

WouterJ commented on Doctrine/Phpcr/SimpleUrlInformationGuesser.php:
```diff
@@ -0,0 +1,89 @@
+        $seoMetadata = $this->seoPresentation->getSeoMetadata($object);
+        if (null !== $seoMetadata->getTitle()) {
+            $urlInformation->setLabel($seoMetadata->getTitle());
+            return $urlInformation;
```
shouldn't it always be returned?

WouterJ replied:
hmm, no, it should never return anything :)

ElectricMaxxx replied:
Exactly. Doing every change on the reference passed as an parameter.

WouterJ commented on Doctrine/Phpcr/SimpleUrlInformationGuesser.php:
```diff
@@ -0,0 +1,89 @@
+ *
+ * @author Maximilian Berghoff <Maximilian.Berghoff@gmx.de>
+ */
+class SimpleUrlInformationGuesser implements UrlInformationGuesserInterface
```
I would split this guesser in 2 guessers (which you can then also give more descriptive names):

* `DefaultValueGuesser` (which should come at last, filling all open values with default values)
* `ObjectGuesser` (which executes the extractors and routing generators, etc.)

I also think this does not belong in the `Doctrine\Phpcr` namespace, as nothing in this class requires PHPCR-ODM stuff.

WouterJ commented on Doctrine/Phpcr/SimpleUrlInformationGuesser.php:
```diff
@@ -0,0 +1,89 @@
+     */
+    public function guessValues(UrlInformation $urlInformation, $object, $sitemap = 'default')
+    {
+        $urlInformation->setLocation($this->router->generate($object, array(), true));
```
we should add some checks first, to see if the value isn't already guessed by another guesser and if this guesser should overrule that value.

Actually, I think we should use a real guess implementation with a confidence. E.g:
````php
$values = array();

$values[] = new LocationGuess(
    $this->router->generate($object, array(), true),
    Guess::VERY_HIGH_CONFIDENCE
);

$values[] = new ChangeFrequencyGuess(
    $this->defaultChangeFrequency,
    Guess::LOW_CONFIDENCE
);

// ...
````

The value with the highest confidence is then set in the UrlInformation instance. Just like https://github.com/symfony/Form/blob/master/Guess/Guess.php

ElectricMaxxx replied:
you are completely right, was just my way of extracting the the lines into their destinations.

ElectricMaxxx replied:
That is a slightly different solution to the priority thing on the attributes, right?

dbu replied:
yeah, this would be the all-out complex solution. it is more powerful and puts less logic into single guessers, as they would not need to check if a field is set already or not. but it adds yet more things. but we could go that way. not sure if there is any reusable guesser infrastructure we could take or whether we would need to completely implement things here.

ElectricMaxxx replied:
@WouterJ got a little dependency problem with that. 

Normaly we try to implement a default behavior and some extra stuff for stuff that dependens on phpcr. For example when phpcr is chosen the `PublicWorkflowUrlInformationVoter` starts working and takes care that public content is shown on sitemap only.
But what should i do for your suggestion of guessers? There are some values that are easy: `label` (SeoMetdata), but the router needs phpcr to return a valid url based on the content object - null would be a nice url.
So what should i do? One guesser, which works when phpcr is chosen? but how would you customize that? I can't work with the chain filled by tagged services, as each one need its own value.

dbu replied:
in the interest of getting something done that works, rather than build a huge thing:

could we simply build this directly as a guesser and not use any chain for the first iteration? as long as we follow the interfaces and allow to configure the service, adding a chain later on should be possible without BC break.

ElectricMaxxx replied:
So you would suggest to have just ONE guesser, which Is able to extract the values and implements an interface, Right? So i would remove the chain and tagging stuff for the guesser. We wouldn't break anything by replacing the chain By @wouterJ 's Suggestion or something else later, Right? Sounds easy. I like your Plan.

dbu replied:
yes, that is my suggestion.

building a full guesser infrastructure sounds huge for just this feature. actually makes me wonder if the symfony form type guesser infrastructure could be extracted into its own component to make it reusable. @webmozart what do you think about that? could this be something for symfony3, building a generic guesser component?

ElectricMaxxx replied:
Yea @dbu i think this would be a nice idea. I would need that guesser component in an actual project in my Company too. 

I will try to finish that PR tonight.

ElectricMaxxx replied:
But @dbu would have a dependecy confusion again:

```
        $urlInformation->setLocation($this->router->generate($object, array(), true));
        $urlInformation->setChangeFrequency($this->defaultChangeFrequency);

        if ($this->alternateLocaleProvider) {
            $collection = $this->alternateLocaleProvider->createForContent($object);
            $urlInformation->setAlternateLocales($collection->toArray());
        }

        $seoMetadata = $this->seoPresentation->getSeoMetadata($object);
        if (null !== $seoMetadata->getTitle()) {
            $urlInformation->setLabel($seoMetadata->getTitle());
            return $urlInformation;
        }
``` 

would be the current code for a simple global guesser for the url information. But there are at least 2 properties, which are created by a service which depends on `PHPCR`: location (router won't work) and alternate locale. Not to set the alternate locale for a non existing service would be ok, but the location? 
At the end the complete guesser depends on `PHCR`, but we won't be able to skip the guesser itself inside of the sitemap controller, as there would be no content at all. So the controller itself depends on `PHPCR`.

ElectricMaxxx commented:
Yea the base structure is prepared now. Just need to re-implement the publich workflow checker as a voter now to get the former tests become green.

Would need some help with naming, but will clear them up a little bit first. 
Any hints for possible voters and guessers i should implement as default?

dbu commented on Controller/SitemapController.php:
```diff
@@ -64,9 +67,15 @@ public function __construct(
      * @return Response
      */
-    public function indexAction($_format)
+    public function indexAction($_format, $sitemap)
```
should we do `$sitempa =  'default'`?

dbu commented on Controller/SitemapController.php:
```diff
@@ -64,9 +67,15 @@ public function __construct(
      *
```
missing phpdoc for the sitemap parameter

dbu commented on DependencyInjection/Compiler/RegisterUrlInformationProviderPass.php:
```diff
@@ -32,13 +32,34 @@ class RegisterUrlInformationProviderPass implements CompilerPassInterface
     public function process(ContainerBuilder $container)
     {
+        $this->processTagsForService(
+            'cmf_seo.sitemap.document_provider',
```
why "document" - i think this should be called "object" here, its not only about phpcr-odm.

dbu commented on DependencyInjection/Compiler/RegisterUrlInformationProviderPass.php:
```diff
@@ -32,13 +32,34 @@ class RegisterUrlInformationProviderPass implements CompilerPassInterface
+        );
+
+        $this->processTagsForService(
+            'cmf_seo.sitemap.guesser_provider',
```
should this not just be "guesser" instead of "guesser_provider"?

dbu commented:
wow, this got pretty big again :-)

i suggest we simplify to the following:

* Sitemap\Provider is a class with no interface, used by the controller. There is one provider per named sitemap. Its only api method is getSitemap() without argument.
* Sitemap\LoaderInterface has method load() that is supposed to return UrlInformation objects or things that can be converted to UrlInformation objects. The SitemapProvider accepts more than one loader and implements the chaining.
* Sitemap\VoterInterface has a method vote() to decide whether a thing returned by the loader should go into the sitemap.
* Sitemap\GuesserInterface has a method guess to enrich a UrlInformation object from the data. there can be several guessers. A guesser SHOULD NOT update a field that is already set. the priorities are thus handled by registering the guessers. I don't see how we could go into more detail here (confidence) without become even more complicated which would seem over the top for this usecase imho. if you really think its worth it, the approach wouter proposed for adding guesses seems right to me.

loaders and guessers are tagged sitemap.loader and sitemap.guesser . you can specify an extra attribute on the tag called sitemap=xxx to limit them to specific sitemaps. that attribute value can be comma separated. they also have a priority attribute.

as for implementations, there is a phpcr specific loader, the published voter tying in the CoreBundle publish workflow and guessers using the seo extractor and a default value guesser.

i want to propose not to pass sitemap names around at all, so you control things by setting up different services on the different SitemapProvider instances. but i am not 100% sure if that is not becoming inconvenient for loaders or voters. having only one provider would complicate things further however, and having both leaves too many different ways to do the same, imho.

wdyt?

ElectricMaxxx commented:
The names sounds sane to me. What i do not understand is the comma separated attribute. You whant something like `sitemap="1,my-sitemap"` instead of `priority="1" sitemap"my-sitemap"`?

I would use the different sitemap approach in all cases, so it will be easy for one (no attributes) and yea for more than one sitemap it gonna be hard, but a page with more than one sitemap will be complicated at all.

I think one provider per sitemap gonna be hard to, cause who wants to inject the right one into the controller? I think have the provider stuff out of the controller is cool, but i don't really wanna have some hard configuration-hacking to have the matching provider inside of the controller.

I will clean up the names now. So we will get a better view later.

WouterJ commented:
@ElectricMaxxx David menas `sitemap="sitemap1, sitemap2, ..." priority="99"`

dbu commented:
yep, exactly thats what i meant with comma separating.

you inject a configuration per sitemap name anyways, for the templates. the provider should just be one field in that configuration. i think this is the most elegant aproach to tweak when to use which voters or providers (or guessers, if they are different). it means that we need to copy the service definition for each sitemap name, as per the configuration of the templates. we should also validate the tag to only reference existing sitemaps

ElectricMaxxx commented:
Cleared the names of classes, services and tags. So let the big refactoring began. Tests are green again. So the parts are working as before.

ElectricMaxxx commented:
So ...
What is missing now?

* multiple sitemap names: `sitemap="name1, name2,.."
*  beforeValidation callbacks for configuration

dbu commented on Controller/SitemapController.php:
```diff
@@ -25,48 +26,57 @@
         $this->templating = $templating;
-        $this->templates = $templates;
+        $this->configurations = $configurations;
+        $this->informationProvider = $informationProvider;
```
can we call this variable `sitemapProvider`?

dbu commented on DependencyInjection/CmfSeoExtension.php:
```diff
@@ -187,26 +188,30 @@ private function loadPhpcr($config, XmlFileLoader $loader, ContainerBuilder $con
+                        'setAlternateLocaleProvider',
+                        array($alternateLocaleProviderDefinition)
+                    );
+            } catch(InvalidArgumentException $e) {
```
can we remove this try-catch again?

ElectricMaxxx replied:
no we can't. 
There are situations, where the service `cmf_seo.sitemap.guesser` isn't available - alternate locale actived, sitmap not. That would cause an exception for the `findDefintion()` method.

dbu replied:
but then we should do $container->hasDefinition or whatever its called to check, if we can't even set a flag in the class to know whether the guesser should be available or not. silently discarding an exception is quite risky, and should not be needed for this case.

ElectricMaxxx replied:
thats a thing @WouterJ suggested to do: use `findDefintion()` at all don't use `hasDefintion()`.

WouterJ replied:
I would propose to use `if ($container->has('...')) { $something = $container->findDefinition('...'); }`

dbu commented on Doctrine/Phpcr/PublicWorkflowUrlInformationVoter.php:
```diff
@@ -0,0 +1,44 @@
+     */
+    public function exposeOnSitemap($content, $sitemap = 'default')
+    {
+        return $this->publishWorkflowChecker->isGranted(array(PublishWorkflowChecker::VIEW_ATTRIBUTE), $content);
```
no need to wrap the attribute in an array, the checker does that for you

dbu commented on Doctrine/Phpcr/PublicWorkflowUrlInformationVoter.php:
```diff
@@ -0,0 +1,44 @@
+     * @param string $sitemap
+     *
+     * @return bool
+     */
```
use `{@inheritDoc}` here

dbu commented on Doctrine/Phpcr/PublicWorkflowUrlInformationVoter.php:
```diff
@@ -0,0 +1,44 @@
+/**
+ * @author Maximilian Berghoff <Maximilian.Berghoff@mayflower.de>
+ */
+class PublicWorkflowUrlInformationVoter implements VoterInterface
```
lets call this PublishWorkflowUrlInformationVoter and not Public..

dbu commented on Doctrine/Phpcr/PublicWorkflowUrlInformationVoter.php:
```diff
@@ -0,0 +1,44 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Cmf\Bundle\SeoBundle\Doctrine\Phpcr;
```
this voter is not specific to phpcr or doctrine in general.

dbu commented on Sitemap/GuesserInterface.php:
```diff
@@ -0,0 +1,29 @@
+interface GuesserInterface
+{
+    /**
+     * Guesses url information values from a given object.
```
lets also say that this method updates $urlInformation with new values.

dbu commented:
this needs a rebase.
travis is very unhappy about some BC break with sonata blocks - we probably need to restrict what version we are compatible with.

dbu commented:
i will be away for another week, so wont have time very soon to do something. if nobody gets to it before, i can then try to contribute the configuration thing for the default sitemap configuration. 
what do you mean by the "multiple sitemap names"?

ElectricMaxxx commented:
Any ideas what goes wrong on travis?

dbu commented on DependencyInjection/CmfSeoExtension.php:
```diff
@@ -208,23 +211,25 @@ private function loadAlternateLocaleProvider($config, ContainerBuilder $containe
-                    array($alternateLocaleProviderDefinition)
-                )
-            ;
+        if ($alternateLocaleProvider) {
```
we already returned above if $alternateLocaleProvider is not set, so this check can be removed and indention be reverted.

dbu commented on Sitemap/AbstractChain.php:
```diff
@@ -0,0 +1,56 @@
+namespace Symfony\Cmf\Bundle\SeoBundle\Sitemap;
+
+/**
+ * @author Maximilian Berghoff <Maximilian.Berghoff@gmx.de>
```
can you please add a phpdoc here what this is?

dbu commented on Sitemap/Guesser.php:
```diff
@@ -0,0 +1,88 @@
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * A chain to register all url information guesser by its priority and preferred sitemap.
```
i don't see a chain in here. copy-paste?

dbu commented on Sitemap/GuesserInterface.php:
```diff
@@ -0,0 +1,29 @@
+use Symfony\Cmf\Bundle\SeoBundle\Model\UrlInformation;
+
+/**
+ * @author Maximilian Berghoff <Maximilian.Berghoff@gmx.de>
```
can you please add a short explanation what this interface does?

dbu commented on Sitemap/AbstractChain.php:
```diff
@@ -0,0 +1,56 @@
+     *
+     * @return array
+     */
+    protected function getSortedItemsBySitemap($sitemap)
```
i think this should be called "getSortedItemsForSitemap". by sitemap made me think that the items are sorted by sitemap. but we only ever get items for one specific sitemap

dbu commented on Sitemap/AbstractChain.php:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * @param $sitemap
+     *
+     * @return array
```
can you explain a bit what this does?

dbu commented on Sitemap/AbstractChain.php:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * The list of guessers with sitemap name and priorities as key.
+     *
+     * @var
```
this is an array of objects.

dbu commented on Sitemap/Voter.php:
```diff
@@ -0,0 +1,27 @@
+    /**
+     * {@inheritDocs}
+     *
+     * First comes first - The first negative voter decides.
```
in terms of symfony security, this would be called "unanimous": every voter can veto a content.

dbu commented on Sitemap/Voter.php:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @author Maximilian Berghoff <Maximilian.Berghoff@mayflower.de>
+ */
+class Voter extends AbstractChain implements VoterInterface
```
don't remember if we discussed this before, but could we remove the whole voting implementation and use our own instance of `AccessDecisionManager` with strategy `unanimous` and have the voters implement the symfony security `VoterInterface`?

i think the security component does a good job at modeling the whole voting process.

dbu replied:
the publish workflow in core bundle is doing this successfully, could be an example to look at how to do this.

ElectricMaxxx replied:
uhhh, ohhh this would make my abstraction for the chain useless, as there will be one chain afterwards.

dbu replied:
which would mean less code to maintain ;-)

ElectricMaxxx replied:
@dbu i can't really follow your ideas with the `VoterInterface`. The `attribute` can be like:
```
[ sitemap => 'sitemap1, sitemap2, ..']
```
The object is the document/content to vote for, but what should i use as a token? I have got no security context here, is there an empty token (new) that i should use?

dbu replied:
lets drop my idea :-/

ElectricMaxxx replied:
So i would just fix tests and then merge?

Am 16.07.2015 um 08:55 schrieb David Buchmann:
>> @@ -0,0 +1,27 @@
>> +<?php
>> +
>> +namespace Symfony\Cmf\Bundle\SeoBundle\Sitemap;
>> +
>> +/**
>> + * A chain as a decorator for all tagged voter services.
>> + *
>> + * @author Maximilian Berghoff <Maximilian.Berghoff@mayflower.de>
>> + */
>> +class Voter extends AbstractChain implements VoterInterface
> lets drop my idea :-/
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/symfony-cmf/SeoBundle/pull/225/files#r34760191

dbu commented:
i think what we saw on travis is related to changes in the testing component and phpcr-odm. i just restarted to see if that helps.

WouterJ commented:
The tests are failing because PHP code coverage needs to be a stable version in order to support HHVM. Add `"phpunit/php-code-coverage": "@stable"` to `require-dev` to solve the problem.

ElectricMaxxx commented:
@dbu i gave you some help:
- rebased the branch
- reverted the voter changes i did
- test should be green in general

but there is an method signature fail in `\Sonata\AdminBundle\Block\AdminListBlockService` localy. I fixed that localy to get the tests green, but we should be aware of that.

dbu commented:
thanks. i started some refactoring, will do a PR against this PR soonish.

ElectricMaxxx commented:
And please don't forget the task with the configuration: singel/multiple sitemaps. Thanks a Lot.

ElectricMaxxx commented:
I will hopefully have some Time to review the PR tonight.

ElectricMaxxx commented on DependencyInjection/Configuration.php:
```diff
@@ -102,18 +102,31 @@ public function getConfigTreeBuilder()
                     ->end()
                 ->end()
                 ->arrayNode('sitemap')
```
@dbu didn't we want to change/extend this? You wanted to use single configurations too, right?

dbu replied:
ah right, now i remember what we said. will give that a try tonight.

dbu commented:
okay, now we have more stuff. what was the task with the multiple sitemaps configuration again? is that covered or are there more things to consider?

was it only the failing tests or is there other things that remain to be done?

ElectricMaxxx commented:
I mean your own commit:

https://github.com/symfony-cmf/SeoBundle/pull/225#discussion_r24973222
:-)

The tests had been fixed (before your PR)

ElectricMaxxx commented:
yea the tests are still green, so the configuration thing is missing only.

ElectricMaxxx commented:
rebased and squashed, merge when green?

dbu commented:
:+1: - please remove the [WIP] from the title :-)

dbu commented:
\o/
